### DISCUSSION TITLE
Carbon's build system and package manager

### DISCUSSION THREAD

**Discussion by felixf4xu** [carbon-language/carbon-lang#3669] [General]
Hi,

I figured out how to compile this repo locally on my WSL2. It's not a surprise to see this project is built by `bazel` (which is need JDK) like other google repos. I wasted some time with bazel.

But it reminded me on another topic, what will be the **build tool/package manager** for Carbon? 

rust has cargo and C++ failed at that part. I hope one day carbon will fix it as the second mistake of c++ ( the first mistake of c++ is the template thing)

I checked the current carbon, it seems that `carbon` itself will the be build tool, then we only need the package manager :-)

```sh
carbon --help
Carbon Language toolchain -- version 0.0.0

This is the unified Carbon Language toolchain driver. Its subcommands provide all of the core behavior of the toolchain, including compilation, linking, and developer tools. Each of these has its own subcommand, and you can pass a specific subcommand to the `help` subcommand to get details about its usage.

Usage:
  carbon [-v] compile [OPTIONS] <FILE>...

Subcommands:
  compile
          Compile Carbon source code.

          This subcommand runs the Carbon compiler over input source code, checking it for errors and producing the requested output.

          Error messages are written to the standard error stream.

          Different phases of the compiler can be selected to run, and intermediate state can be written to standard output as these phases progress.
```

**Comment by Mustafif** [comment]
That's definitely true, C++ didn't have an official package manager, and with my experience using package managers like Cargo, npm, and even Zig's package manager, I feel like the way Cargo was done is so far my favourite, and could be a good base of inspiration (using github to be the package management is really not ideal *cough cough Go*).

**Reply by red-co** [reply]
"Contact me if you want to use this name!".  

Cargo, the package manager namespace. Squatting. Don't show up in carbon. I prefer go. It actually provides URLs so you can go directly to where the source code is. Rust's library names are as confusing as cpp, cough, cough. The author name or source code address should come before the project name.    
   
Crate.io:: "Current policy is to not take action".  Lol

**Reply by josh11b** [reply]
Please speak about other programming languages in a respectful manner.

**Comment by jonmeow** [comment]
A package manager is planned. The current priority is providing a compiler for language evaluation, for the [0.1 milestone](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/milestones.md). A package manager will probably be started before the 1.0 milestone, but there's not a specific milestone for it right now. cargo's sure to be an inspiration, but I don't think there's been much design discussion.

Although we use bazel to build carbon, the intent is to make the carbon compiler work more broadly with other build systems (e.g. cmake), fitting in similar to clang. The [language tools and ecosystem](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#language-tools-and-ecosystem) goal is related here and may also provide some helpful info.

**Reply by felixf4xu** [reply]
thanks for the link, it's clearly stated.

**Comment by Jark5455** [comment]
How would a package manager distribute packages from a centralized index? I believe currently cargo builds each package from source, but for non-open source dependencies this may be a problem. If we were to only distribute the binaries (assuming that carbon is intended to support the wide range of platforms that C++ currently supports) carrying the binaries for every single combination of OS, CPU architecture, kernel, etc seems very expensive and unfeasable.
